### PR TITLE
[SPARK-50526][SS] Add store encoding format conf into offset log and block non supported stateful operators from using avro

### PR DESCRIPTION
**Pull request by anishshri-db**
### What changes were proposed in this pull request?
Add store encoding format conf into offset log and block non supported stateful operators from using avro

### Why are the changes needed?
Changes are needed to ensure that encoding format info is stored in offset log and that trying to use avro in a query that has stateful operators that don't support this encoding format yet will result in query failure.

### Does this PR introduce _any_ user-facing change?
Yes

### How was this patch tested?
Added unit tests

```
[info] Run completed in 2 seconds, 547 milliseconds.
[info] Total number of tests run: 6
[info] Suites: completed 1, aborted 0
[info] Tests: succeeded 6, failed 0, canceled 0, ignored 0, pending 0
[info] All tests passed.
[success] Total time: 13 s, completed Dec 9, 2024, 6:15:25 PM
```

### Was this patch authored or co-authored using generative AI tooling?
No

### CODE REVIEW

**Comment by anishshri-db** [comment]
cc - @ericm-db @HeartSaVioR - PTAL, thx !

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingDeduplicationSuite.scala [hunk 1]
```diff
@@ -574,6 +574,24 @@ class StreamingDeduplicationSuite extends StateStoreMetricsTest {
       matchPVals = true
     )
   }
+
+  test("test that avro encoding is not supported") {
```
Comment (ericm-db) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingDeduplicationSuite.scala:578]: Can we add a test case that uses TWS and another stateful operator?

**Review by ericm-db — APPROVED** [review]
LGTM pending comment

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingDeduplicationSuite.scala (hunk 1 above)
Comment (anishshri-db) [reply sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingDeduplicationSuite.scala:578]: Done

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingAggregationSuite.scala [hunk 2]
```diff
@@ -874,6 +874,34 @@ class StreamingAggregationSuite extends StateStoreMetricsTest with Assertions {
...
+      inputData.toDF()
+        .groupBy($"value")
+        .agg(count("*"))
+        .as[(Int, Long)]
+
+    val ex = intercept[Exception] {
+      withSQLConf(SQLConf.STREAMING_STATE_STORE_ENCODING_FORMAT.key -> "avro") {
+        testStream(aggregated, Update)(
+          AddData(inputData, 3),
+          CheckLastBatch((3, 1)),
```
Comment (HeartSaVioR) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingAggregationSuite.scala:890]: nit: ditto, no need to have a full code

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingDeduplicationSuite.scala [hunk 3]
```diff
@@ -574,6 +574,24 @@ class StreamingDeduplicationSuite extends StateStoreMetricsTest {
...
+    val inputData = MemoryStream[String]
+    val result = inputData.toDS().dropDuplicates()
+
+    val ex = intercept[Exception] {
+      withSQLConf(SQLConf.STREAMING_STATE_STORE_ENCODING_FORMAT.key -> "avro") {
+        testStream(result, Append)(
+          AddData(inputData, "a"),
+          CheckLastBatch("a"),
```
Comment (HeartSaVioR) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingDeduplicationSuite.scala:586]: ditto, no need to have a full code

**Review by HeartSaVioR — COMMENTED** [review]
Shall we add the test for checking the default value? We can use the traditional approach with Spark 3.5 checkpoint.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationChecker.scala [hunk 4]
```diff
@@ -140,6 +142,37 @@ object UnsupportedOperationChecker extends Logging {
...
+  }
+
+  // Rule to check that avro encoding format is not supported in case any
+  // non-transformWithState stateful streaming operators are present in the query.
+  def checkSupportedStoreEncodingFormats(plan: LogicalPlan): Unit = {
+    val storeEncodingFormat = SQLConf.get.stateStoreEncodingFormat
+    if (storeEncodingFormat.toLowerCase(Locale.ROOT) == "avro") {
+      plan.foreach { subPlan =>
+        if (!checkAvroSupportForStatefulOperator(subPlan)) {
+          val errorMsg = "State store encoding format as avro is not supported for " +
```
[the remark this answers is not in the document]
Comment (anishshri-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationChecker.scala:169]: Done

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingDeduplicationSuite.scala (hunk 3 above)
Comment (anishshri-db) [reply sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingDeduplicationSuite.scala]: Done

**Comment by anishshri-db** [comment]
> Shall we add the test for checking the default value? We can use the traditional approach with Spark 3.5 checkpoint.

Done - added the test

**Review by HeartSaVioR — APPROVED** [review]
+1

**Comment by HeartSaVioR** [comment]
Thanks! Merging to master.

**Pull request closed without merging**